File: backend/platform/src/main/java/ru/abondin/hreasy/platform/auth/DbAuthoritiesPopulator.java
package ru.abondin.hreasy.platform.auth;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ldap.core.DirContextOperations;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.ldap.userdetails.LdapAuthoritiesPopulator;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import ru.abondin.hreasy.platform.sec.SecurityService;

import java.util.Collection;

/**
 * Setup authorities for user from cookie or database
 */
@Component
@Slf4j
@RequiredArgsConstructor
public class DbAuthoritiesPopulator implements LdapAuthoritiesPopulator {
    private final SecurityService securityService;

    @Override
    public Collection<? extends GrantedAuthority> getGrantedAuthorities(DirContextOperations userData,
                                                                        String username) {
        return getGrantedAuthoritiesBlocked(username);
    }

    private Collection<? extends GrantedAuthority> getGrantedAuthoritiesBlocked(String username) {
        return securityService
                .getPermissions(username)
                .map(perm -> new SimpleGrantedAuthority(perm))
                .collectList()
                //FIXME blocking code to support not reactive LdapAuthoritiesPopulator
                .blockOptional().get();
    }

    public Flux<GrantedAuthority> getGrantedAuthorities(String username) {
        return securityService
                .getPermissions(username)
                .map(perm -> new SimpleGrantedAuthority(perm));
    }
}

File: platform/src/main/kotlin/ru/abondin/hreasy/platform/sec/DbAuthoritiesPopulator.kt
package ru.abondin.hreasy.platform.sec

import org.springframework.ldap.core.DirContextOperations
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.ldap.userdetails.LdapAuthoritiesPopulator
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux

/**
 * Setup authorities for user from cookie or database
 */
@Component
class DbAuthoritiesPopulator(
        private val securityService: SecurityService
) : LdapAuthoritiesPopulator {

    override fun getGrantedAuthorities(userData: DirContextOperations?, username: String): MutableCollection<out GrantedAuthority> {
        return getGrantedAuthoritiesBlocked(username);
    }

    private fun getGrantedAuthoritiesBlocked(username: String): MutableCollection<out GrantedAuthority> {
        return securityService
                .getPermissions(AuthHandler.emailFromUsername(username))
                .map { perm -> SimpleGrantedAuthority(perm) }
                .collectList()
                //FIXME blocking code to support not reactive LdapAuthoritiesPopulator
                .blockOptional().get();
    }

    fun getGrantedAuthorities(username: String): Flux<out GrantedAuthority> {
        return securityService
                .getPermissions(AuthHandler.emailFromUsername(username))
                .map { perm -> SimpleGrantedAuthority(perm) };
    }
}